#### Build a Regression model that predicts the profit of startup on basis of investment they have made in RND(research and Development)

In [21]:
import warnings
warnings.filterwarnings("ignore")

In [22]:
#Step 1 : Data gathering
path=r"https://raw.githubusercontent.com/sindhura-nk/Datasets/refs/heads/main/50_Startups.csv"
import pandas as pd
df=pd.read_csv(path)
df.head()

,RND,ADMIN,MKT,STATE,PROFIT
0,165349.20,136897.80,471784.10,New York,192261.83
1,162597.70,151377.59,443898.53,California,191792.06
2,153441.51,101145.55,407934.54,Florida,191050.39
3,144372.41,118671.85,383199.62,New York,182901.99
4,142107.34,91391.77,366168.42,Florida,166187.94


Perform basic data quality check

In [23]:
df.shape

(50, 5)

In [24]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50 entries, 0 to 49
Data columns (total 5 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   RND     50 non-null     float64
 1   ADMIN   50 non-null     float64
 2   MKT     50 non-null     float64
 3   STATE   50 non-null     object 
 4   PROFIT  50 non-null     float64
dtypes: float64(4), object(1)
memory usage: 2.1+ KB


In [25]:
## separate x and y features
#x: RND 
#y: profit

In [26]:
X = df[['RND']]
Y = df[['PROFIT']]

In [27]:
X.head()

,RND
0,165349.20
1,162597.70
2,153441.51
3,144372.41
4,142107.34


In [28]:
Y.head()

,PROFIT
0,192261.83
1,191792.06
2,191050.39
3,182901.99
4,166187.94


##### Model Building
Profit_Predicted = B0 + B1.RND

In [29]:

from sklearn.linear_model import LinearRegression
model = LinearRegression()
model.fit(X,Y)

LinearRegression()

In [30]:
model.coef_

array([[0.85429137]])

In [31]:
model.intercept_

array([49032.89914125])

##### Profit_predicted = 49032.9 + (0.85)*RND
If RND is 0, the profit predicted will be 49032
If RND increases by 1000, you will observe that Profit predicted will increase by 850.

In [32]:
X.head()

,RND
0,165349.20
1,162597.70
2,153441.51
3,144372.41
4,142107.34


In [33]:
ypreds = model.predict(X)
ypreds[:10]

array([[190289.29389289],
       [187938.71118575],
       [180116.65707807],
       [172369.00320589],
       [170433.97345032],
       [161694.19683741],
       [164033.72501421],
       [160345.46724972],
       [152011.33380847],
       [154396.82286103]])

In [34]:
Y.head(10)

,PROFIT
0,192261.83
1,191792.06
2,191050.39
3,182901.99
4,166187.94
5,156991.12
6,156122.51
7,155752.60
8,152211.77
9,149759.96


Evaluate the Model Performance

In [35]:
from sklearn.metrics import mean_absolute_error,mean_squared_error,r2_score

mse = mean_squared_error(Y,ypreds)
mae = mean_absolute_error(Y,ypreds)
rmse = mse**(1/2)
r2 = r2_score(Y,ypreds)

print("----------------------")
print(f"Mean Squared Error: {mse:.2f}")
print(f"Root Mean Squared Error: {rmse:.2f}")
print(f"Mean Absolute Error: {mae:.2f}")
print(f"R squared error or R2 score: {r2*100:.2f}%")
print("----------------------")

----------------------
Mean Squared Error: 85120931.33
Root Mean Squared Error: 9226.10
Mean Absolute Error: 6910.98
R squared error or R2 score: 94.65%
----------------------


Above model is a good model as the r2 score is above 0.8. This model can be finalised for deployment
Save the model

In [36]:
model

LinearRegression()


Joblib: Joblib is an open-source Python library designed for efficient serialization and parallel computing. It is widely used in data science and machine learning for saving and loading large Python objects such as NumPy arrays and trained models, particularly in combination with scikit-learn.

In [37]:
import joblib
joblib.dump(model,"model.joblib")

['model.joblib']

In [38]:
m1 = joblib.load("model.joblib")


In [39]:
rnd_data=[185000,145000,190000]


In [40]:
profit_pred = m1.predict([rnd_data])

ValueError: X has 3 features, but LinearRegression is expecting 1 features as input.

In [41]:
profit_pred = m1.predict([rnd_data])
print(profit_pred)

ValueError: X has 3 features, but LinearRegression is expecting 1 features as input.

In [42]:
X.head()

,RND
0,165349.20
1,162597.70
2,153441.51
3,144372.41
4,142107.34


In [43]:
from warnings import filterwarnings
filterwarnings("ignore")

In [44]:
for i in rnd_data:
    profit_pred = m1.predict([[i]])
    print(f"RND: {i} --> Predicted Profit: {profit_pred[0][0]:.2f}")
    

RND: 185000 --> Predicted Profit: 207076.80
RND: 145000 --> Predicted Profit: 172905.15
RND: 190000 --> Predicted Profit: 211348.26
